src/WebDriverTunnel.ts

```diff
@@ -0,0 +1,301 @@
+  // session.  If the executable does not write to stdout, the tunnel will
+  // wait this amount of time in milliseoncds before requesting a new session.
+  // Default is 1000ms (1 second).
+  startupWait?: number;
+}
+
+/**
+ * A WebDriver tunnel that does NOT depend on Java. This tunnel downloads
+ * WebDriver server executables for each supported browsers.
+ * See webDriverConstructors in webDriverConfig.ts for a list of supported
+ * browsers.
+ *
+ * Use WebDriverTunnelOptions to configure the tunnel to use different
+ * versions of the WebDriver servers or even non-standard implementations.
+ *
+ * * `MicrosoftEdge`
+ * * `chrome`
+ * * `firefox`
+ * * `internet explorer`
+ * * `safari`
+ */
+export default class WebDriverTunnel extends Tunnel {
+  private readonly webDriverConfigs: { [key: string]: WebDriverConfig };
+  private webDriverProxy: WebDriverProxy;
+  private lastUsedPort: number;
+  private driverDirectory!: string;
+  private drivers!: WebDriverDownloadOptions[];
+  private maxConnectAttempts!: number;
+  private startupWait!: number;
+
+  constructor(options: WebDriverTunnelOptions) {
+    super(
+      Object.assign(
+        {
+          driverDirectory: 'webdriver_tunnel',
+          maxConnectAttempts: 3,
+          startupWait: 1000
+        },
+        options
+      )
+    );
+
+    const { pathname, port, maxConnectAttempts } = this;
+
+    this.lastUsedPort = Number(this.port);
+
+    // Instantiate the WebDriver config objects.  Pass the tunnel options
+    // to each one as appropriate.
+    const webDriverConfigs: {
+      [key: string]: WebDriverConfig;
+    } = (this.webDriverConfigs = {});
+
+    Object.keys(webDriverConstructors).forEach(browserName => {
+      const Constructor = webDriverConstructors[browserName];
+      webDriverConfigs[browserName] = new Constructor(
+        this.webDriverOptionsFactory(browserName) || {}
+      );
+    });
+
+    this.webDriverProxy = new WebDriverProxy({
+      port: Number(port),
+      path: pathname,
+      maxConnectAttempts: maxConnectAttempts,
+      webDriverChildFactory: browserName => {
+        return this.createWebDriverChild(browserName);
+      }
+    });
+  }
+
+  /**
+   * Find settings in the drivers property that correspond to the given browser name.
+   * This code will use the first object it finds in the drivers array that matches the
+   * browser name and ignore any others that match.
+   * @param browserName
+   */
+  private webDriverOptionsFactory(browserName: string): WebDriverConfigOptions {
+    const { drivers } = this;
+    let options: WebDriverConfigOptions = {};
+
+    drivers &&
+      drivers.some(driverOptions => {
+        if (driverOptions.name === browserName) {
+          options = { ...driverOptions };
+          return true;
+        }
+        return false;
+      });
+
+    return options;
+  }
+
+  get directory() {
+    return join(this.basePath, this.driverDirectory);
+  }
+
+  get isDownloaded() {
```

Comment:
It'd be ideal if we could use the same or similar download logic as SeleniumTunnel (like, have SeleniumTunnel and WebDriverTunnel inherit from a shared base class, or put the download management logic in a separate utility module) for this and the `download` method below.

Reply:
After I get all of the browsers working, I was going to compare what I came up with to the Selenium tunnel and try to make them share more.